File: BHAVRP_Python/cujae/inf/citi/om/heuristic/assignment/clustering/hierarchical/upgmc.py
import numpy as np
from typing import List
from hierarchical import Hierarchical
from .....controller.utils.distance_type import DistanceType
from .....problem.input.problem import Problem
from .....problem.input.customer import Customer
from .....problem.input.depot import Depot
from .....problem.input.location import Location
from .....problem.output.solution.solution import Solution
from .....problem.output.solution.cluster import Cluster

class UPGMC(Hierarchical):
    
    def __init__(self):
        super().__init__()
        
    distance_type = DistanceType.Euclidean;
    
    def to_clustering(self) -> Solution:
        solution = Solution()
        
        # Inicializar listas.
        list_id_elements: List[int] = list(Problem.get_problem().get_list_id_elements())
        list_clusters: List[Cluster] = self.initialize_clusters(list_id_elements)
        list_customers_to_assign: List[Customer] = list(Problem.get_problem().get_customers())
        list_depots: List[Depot] = list(Problem.get_problem().get_depots())
        
        total_depots = len(list_depots)
        change = True
        
        # Inicializar matriz de costos.
        cost_matrix: np.ndarray = None
        
        while list_customers_to_assign and total_depots > 0:
            current_depots = len(list_depots)
            current_customers = len(list_customers_to_assign)
            
            if change:
                cost_matrix = Problem.get_problem().fill_cost_matrix(
                    list_customers_to_assign, list_depots, UPGMC.distance_type
                )
                
            # Encontrar la mejor fila y columna.
            row_col_best_all = np.argmin(
                cost_matrix[0:current_customers + current_depots, 0:current_customers]
            )
            pos_row, pos_col = np.unravel_index(row_col_best_all, cost_matrix.shape)
            
            if pos_col < current_customers and pos_row < current_customers:
                id_customer_one = list_customers_to_assign[pos_col].get_id_customer()
                id_customer_two = list_customers_to_assign[pos_row].get_id_customer()

                print(f"ID Customer One: {id_customer_one}")
                print(f"ID Customer Two: {id_customer_two}")
                
                pos_cluster_one = self.find_cluster(id_customer_one, list_clusters)
                pos_cluster_two = self.find_cluster(id_customer_two, list_clusters)
                
                print(f"Position Cluster One: {pos_cluster_one}")
                print(f"Position Cluster Two: {pos_cluster_two}")
                
                if pos_cluster_one != -1 and pos_cluster_two != -1:
                    request_cluster_one = list_clusters[pos_cluster_one].get_request_cluster()
                    request_cluster_two = list_clusters[pos_cluster_two].get_request_cluster()
                    total_request = request_cluster_one + request_cluster_two
                    
                    print(f"Request Cluster One: {request_cluster_one}")
                    print(f"Request Cluster Two: {request_cluster_two}")
                    print(f"Total Request: {total_request}")

                    id_depot_with_mu = self.get_id_cluster_with_mu(list_depots, list_clusters)
                    pos_depot = Problem.get_problem().find_pos_depot(list_depots, id_depot_with_mu)
                    capacity_depot_with_mu = Problem.get_problem().get_total_capacity_by_depot(
                        list_depots[pos_depot]
                    )
                    pos_cluster = self.find_cluster(id_depot_with_mu, list_clusters)
                    request_cluster = list_clusters[pos_cluster].get_request_cluster()
                    
                    print(f"ID Depot with MU: {id_depot_with_mu}")
                    print(f"Position Depot: {pos_depot}")
                    print(f"Capacity Depot with MU: {capacity_depot_with_mu}")
                    print(f"Position Cluster: {pos_cluster}")
                    print(f"Request Cluster: {request_cluster}")

                    if capacity_depot_with_mu >= (request_cluster + total_request):

                        list_clusters[pos_cluster_one].set_request_cluster(total_request)
                        list_clusters[pos_cluster_one].get_items_of_cluster().extend(
                            list_clusters[pos_cluster_two].get_items_of_cluster()
                        )
                        
                        new_location: Location = self.recalculate_centroid(list_clusters[pos_cluster_one])
                        
                        print(f"New Location X: {new_location.get_axis_x()}")
                        print(f"New Location Y: {new_location.get_axis_y()}")

                        pos_customer_one = Problem.get_problem().find_pos_customer(
                            list_customers_to_assign, id_customer_one
                        )
                        list_customers_to_assign[pos_customer_one].set_location_customer(new_location)
                        
                        print(f"New Location X: {new_location.get_axis_x()}")
                        print(f"New Location Y: {new_location.get_axis_y()}")
                        print(f"Position Customer One: {pos_customer_one}")                        

                        list_clusters.pop(pos_cluster_two)
                        pos_customer_two = Problem.get_problem().find_pos_customer(
                                list_customers_to_assign, id_customer_two
                        )
                        list_customers_to_assign.pop(pos_customer_two)
                        
                        print(f"List Clusters: {len(list_clusters)}")
                        print(f"Position Customer Two: {pos_customer_two}")
                        print(f"List Customers to Assign: {len(list_customers_to_assign)}")
                        
                        change = True
                    else:
                        cost_matrix[pos_row, pos_col] = float("inf")
                        change = False

            elif (
                (pos_col < current_customers <= pos_row)
                or (pos_row < current_customers <= pos_col)
            ):
                if pos_col < current_customers:
                    pos_customer_one, pos_depot_matrix = pos_col, pos_row
                else:
                    pos_customer_one, pos_depot_matrix = pos_row, pos_col
                    
                print(f"Position Customer One: {pos_customer_one}")
                print(f"Position Depot Matrix: {pos_depot_matrix}")

                id_customer_one = list_customers_to_assign[pos_customer_one].get_id_customer()
                pos_cluster_one = self.find_cluster(id_customer_one, list_clusters)

                pos_depot = pos_depot_matrix - current_customers
                id_depot = list_depots[pos_depot].get_id_depot()
                capacity_depot = Problem.get_problem().get_total_capacity_by_depot(
                    list_depots[pos_depot]
                )
                pos_cluster = self.find_cluster(id_depot, list_clusters)
                
                print(f"Position Depot: {pos_depot}")
                print(f"ID Depot: {id_depot}")
                print(f"Capacity Depot: {capacity_depot}")
                print(f"Position Cluster: {pos_cluster}")

                if pos_cluster_one != -1 and pos_cluster != -1:
                    request_cluster_one = list_clusters[pos_cluster_one].get_request_cluster()
                    request_cluster = list_clusters[pos_cluster].get_request_cluster()
                    
                    print(f"Request Cluster One: {request_cluster_one}")
                    print(f"Request Cluster: {request_cluster}")

                    if capacity_depot >= request_cluster + request_cluster_one:
                        list_clusters[pos_cluster].set_request_cluster(
                            request_cluster + request_cluster_one
                        )
                        list_clusters[pos_cluster].get_items_of_cluster().extend(
                            list_clusters[pos_cluster_one].get_items_of_cluster()
                        )
                        list_customers_to_assign.pop(
                            Problem.get_problem().find_pos_customer(
                                list_customers_to_assign, id_customer_one
                            )
                        )
                        list_clusters.pop(pos_cluster_one)
                        change = True
                    else:
                        cost_matrix[pos_row, pos_col] = float("inf")
                        change = False

                # Verificar si el depósito está lleno
                if self.is_full_depot(
                    list_clusters, request_cluster, capacity_depot, len(list_customers_to_assign)
                ):
                    pos_cluster = self.find_cluster(id_depot, list_clusters)
                    print(f"Position Cluster: {pos_cluster}")
                    
                    if list_clusters[pos_cluster].get_items_of_cluster():
                        solution.get_clusters().append(list_clusters.pop(pos_cluster))
                    else:
                        list_clusters.pop(pos_cluster)

                    list_depots.pop(pos_depot)
                    total_depots -= 1
                    change = True
                    
        self.finish(list_clusters, solution)
        
        # Verificamos si list_clusters no está vacía.
        if list_clusters:
            for cluster in list_clusters:
                if cluster.get_items_of_cluster():
                    solution.clusters.append(cluster)
        
        return solution
    
    # Método encargado de obtener el id del depósito con mayor capacidad de la lista.
    def get_id_cluster_with_mu(self, depots: List[Depot], clusters: List[Cluster]) -> int:
        id_depot_mu = depots[0].get_id_depot()
        pos_cluster = self.find_cluster(id_depot_mu, clusters)
        request_cluster = clusters[pos_cluster].get_request_cluster()
        max_capacity_depot = Problem.get_problem().get_total_capacity_by_depot(depots[0])
        max_capacity_depot -= request_cluster
        total_depots = len(depots)

        current_capacity_depot = 0

        for i in range(1, total_depots):
            pos_cluster = self.find_cluster(depots[i].get_id_depot(), clusters)
            request_cluster = clusters[pos_cluster].get_request_cluster()
            current_capacity_depot = Problem.get_problem().get_total_capacity_by_depot(depots[i])
            current_capacity_depot -= request_cluster

            if max_capacity_depot < current_capacity_depot:
                max_capacity_depot = current_capacity_depot
                id_depot_mu = depots[i].get_id_depot()

        return id_depot_mu
    
    # Verifica si el depósito está lleno.
    def is_full_depot(
        self, 
        clusters: List[Cluster], 
        request_cluster: float, 
        capacity_depot: float, 
        current_customer: int
    ) -> bool:
        is_full = True

        current_request = capacity_depot - request_cluster

        if current_request > 0:
            i = 0

            while i < len(clusters) and i < current_customer and is_full:
                if clusters[i].get_request_cluster() <= current_request:
                    is_full = False
                else:
                    i += 1

        return is_full
    
    # Finaliza el proceso y agrega elementos no asignados.
    def finish(self, clusters: List[Cluster], solution: Solution):
        pos_element = -1

        i = 0
        while i < len(clusters):
            
            print("\n")
            print(f"ID Cluster: {clusters[i].get_id_cluster()}")
            print(f"Position Cluster: {clusters[i].get_items_of_cluster()}")
            print(f"Request Cluster: {clusters[i].get_request_cluster()}")

            pos_element = Problem.get_problem().find_pos_customer(Problem.get_problem().get_customers(), clusters[i].get_id_cluster())

            if pos_element != -1:
                for item in clusters[i].get_items_of_cluster():
                    solution.get_unassigned_items().append(int(item))

                del clusters[i]  # Remover el cluster

            else:
                i += 1

File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/heuristic/assignment/clustering/hierarchical/UPGMC.java
package cujae.inf.ic.om.heuristic.assignment.clustering.hierarchical;

import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;

import cujae.inf.ic.om.factory.DistanceType;

import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Depot;
import cujae.inf.ic.om.problem.input.Location;
import cujae.inf.ic.om.problem.input.Problem;

import cujae.inf.ic.om.problem.output.solution.Cluster;
import cujae.inf.ic.om.problem.output.solution.Solution;

import cujae.inf.ic.om.matrix.NumericMatrix;
import cujae.inf.ic.om.matrix.RowCol;

public class UPGMC extends Hierarchical {

	public static DistanceType distanceType = DistanceType.Euclidean;

	public UPGMC() {
		super();
	}

	@Override
	public Solution toClustering() {
		Solution solution = new Solution();	

		ArrayList<Integer> listIDElements = Problem.getProblem().getListIDElements();
		ArrayList<Cluster> listClusters = initializeClusters(listIDElements);

		NumericMatrix costMatrix = new NumericMatrix();

		ArrayList<Customer> listCustomersToAssign = new ArrayList<Customer>(Problem.getProblem().getCustomers());
		ArrayList<Depot> listDepots = new ArrayList<Depot>(Problem.getProblem().getDepots());
		
		int totalDepots = listDepots.size();
		
		int currentDepots =  -1;
		int currentCustomers = -1;

		int posCol = -1;
		int posRow = -1;

		int idCustomerOne = -1;
		int posCustomerOne = -1;
		int idCustomerTwo = -1;
		int posCustomerTwo = -1;

		int posClusterOne = -1;
		double requestClusterOne = 0.0;
		int posClusterTwo = -1;
		double requestClusterTwo = 0.0;

		int idDepot = -1;
		int posDepot = -1;		
		int posDepotMatrix = -1;
		double capacityDepot = 0.0;

		int posCluster = -1;
		double requestCluster = 0.0;
		double totalRequest = 0.0;

		int idDepotWithMU = -1;
		double capacityDepotWithMU = 0.0;
		RowCol rcBestAll = new RowCol();
		Location newLocation = new Location();
		
		boolean change = true; 

		while(!(listCustomersToAssign.isEmpty()) && (totalDepots > 0))
		{
			currentDepots = listDepots.size();
			currentCustomers = listCustomersToAssign.size();
			
			System.out.println("currentDepots" + currentDepots);
			System.out.println("currentCustomers" + currentCustomers);
			
			//System.out.println("currentCustomers" + InfoProblem.getProblem().getCustomers().size());

			if(change)
			{
				try {
					costMatrix = Problem.getProblem().fillCostMatrix(listCustomersToAssign, listDepots, distanceType);
				} catch (IllegalArgumentException | SecurityException
						| ClassNotFoundException | InstantiationException
						| IllegalAccessException | InvocationTargetException
						| NoSuchMethodException e) {
					e.printStackTrace();
				}
			}
			
			rcBestAll = costMatrix.indexLowerValue(0, 0, (currentCustomers + currentDepots - 1), (currentCustomers - 1));
			
			System.out.println("rcBestAll " + costMatrix.getItem(rcBestAll.getRow(), rcBestAll.getCol()));
			System.out.println("bestAllRow" + rcBestAll.getRow());
			System.out.println("bestAllCol" + rcBestAll.getCol());
			
			posCol = rcBestAll.getCol();
			posRow = rcBestAll.getRow();

			if((posCol < currentCustomers) && (posRow < currentCustomers)) 
			{ 
				posCustomerOne = posCol;
				idCustomerOne = listCustomersToAssign.get(posCustomerOne).getIDCustomer();

				System.out.println("idCustomerOne" + idCustomerOne);
				
				posCustomerTwo = posRow;
				idCustomerTwo = listCustomersToAssign.get(posCustomerTwo).getIDCustomer();				

				System.out.println("idCustomerTwo" + idCustomerTwo);
				
				posClusterOne = findCluster(idCustomerOne, listClusters);	
				posClusterTwo = findCluster(idCustomerTwo, listClusters);
				
				System.out.println("posClusterOne" + posClusterOne);
				System.out.println("posClusterTwo" + posClusterTwo);

				if((posClusterOne != -1) && (posClusterTwo != -1)) 
				{
					requestClusterOne = listClusters.get(posClusterOne).getRequestCluster();
					requestClusterTwo = listClusters.get(posClusterTwo).getRequestCluster();
					totalRequest = requestClusterOne + requestClusterTwo;
					
					System.out.println("requestClusterOne" + requestClusterOne);
					System.out.println("requestClusterTwo" + requestClusterTwo);
					System.out.println("totalRequest" + totalRequest);
					
					idDepotWithMU = getIDClusterWithMU(listDepots, listClusters);
					posDepot = Problem.getProblem().findPosDepot(listDepots, idDepotWithMU);
					capacityDepotWithMU = Problem.getProblem().getTotalCapacityByDepot(listDepots.get(posDepot));
					posCluster = findCluster(idDepotWithMU, listClusters);
					requestCluster = listClusters.get(posCluster).getRequestCluster();

					System.out.println("idDepotWithMU" + idDepotWithMU);
					System.out.println("posDepot" + posDepot);
					System.out.println("capacityDepotWithMU" + capacityDepotWithMU);
					System.out.println("posCluster" + posCluster);
					System.out.println("requestCluster" + requestCluster);
					
					if(capacityDepotWithMU >= (requestCluster + totalRequest)) 
					{
						listClusters.get(posClusterOne).setRequestCluster(totalRequest);

						for (int i = 0; i < listClusters.get(posClusterTwo).getItemsOfCluster().size(); i++)
							listClusters.get(posClusterOne).getItemsOfCluster().add(listClusters.get(posClusterTwo).getItemsOfCluster().get(i));			

						newLocation = recalculateCentroid(listClusters.get(posClusterOne));
						System.out.println("newLocationX" + newLocation.getAxisX());
						System.out.println("newLocationY" + newLocation.getAxisY());
						/*newLocation = recalculatePonderado(listClusters.get(posClusterOne), listClusters.get(posClusterTwo), listCustomersToAssign);
						System.out.println("newLocationX" + newLocation.getAxisX());
						System.out.println("newLocationY" + newLocation.getAxisY());*/
						/*newLocation = recalculateTest(listClusters.get(posClusterOne), listClusters.get(posClusterTwo), listCustomersToAssign);
						System.out.println("newLocationX" + newLocation.getAxisX());
						System.out.println("newLocationY" + newLocation.getAxisY());*/
						posCustomerOne = Problem.getProblem().findPosCustomer(listCustomersToAssign, idCustomerOne);
						listCustomersToAssign.get(posCustomerOne).setLocationCustomer(newLocation);
						
						System.out.println("newLocationX" + newLocation.getAxisX());
						System.out.println("newLocationY" + newLocation.getAxisY());
						System.out.println("posCustomerOne" + posCustomerOne);
						
						listClusters.remove(posClusterTwo);
						posCustomerTwo = Problem.getProblem().findPosCustomer(listCustomersToAssign, idCustomerTwo);
						listCustomersToAssign.remove(posCustomerTwo);
						
						System.out.println("listClusters" + listClusters.size());
						System.out.println("posCustomerTwo" + posCustomerTwo);
						System.out.println("listCustomersToAssign" + listCustomersToAssign.size());
	
						change = true;
					}
					else
					{
						costMatrix.setItem(rcBestAll.getRow(), rcBestAll.getCol(), Double.POSITIVE_INFINITY);
						change = false;
						
						/*System.out.println();
						System.out.println("IDClusterOne" + listClusters.get(posClusterOne).getIDCluster());
						System.out.println("ElementClusterOne" + listClusters.get(posClusterOne).getItemsOfCluster().toString());
						System.out.println("requestClusterOne" + listClusters.get(posClusterTwo).getRequestCluster());
						System.out.println("IDClusterTwo" + listClusters.get(posClusterTwo).getIDCluster());
						System.out.println("ElementClusterTwo" + listClusters.get(posClusterOne).getItemsOfCluster().toString());
						System.out.println("requestClusterTwo" + listClusters.get(posClusterTwo).getRequestCluster());
						*/
					}
				}	
			}
			else
			{				
				if(((posCol < currentCustomers) && (posRow >= currentCustomers)) || ((posRow < currentCustomers) && (posCol >= currentCustomers)))
				{
					if(posCol < currentCustomers)
					{
						posCustomerOne = rcBestAll.getCol();
						posDepotMatrix = rcBestAll.getRow();
					}
					else
					{
						posCustomerOne = rcBestAll.getRow();
						posDepotMatrix = rcBestAll.getCol();
					}
					
					System.out.println("posCustomerOne" + posCustomerOne);
					System.out.println("posDepotMatrix" + posDepotMatrix);

					idCustomerOne = listCustomersToAssign.get(posCustomerOne).getIDCustomer();			
					posClusterOne = findCluster(idCustomerOne, listClusters);
					
					System.out.println("idCustomerOne" + idCustomerOne);
					System.out.println("posClusterOne" + posClusterOne);

					posDepot = (posDepotMatrix - currentCustomers); 
					idDepot = listDepots.get(posDepot).getIDDepot();
					capacityDepot = Problem.getProblem().getTotalCapacityByDepot(listDepots.get(posDepot));
					posCluster = findCluster(idDepot, listClusters);

					System.out.println("posDepot" + posDepot);
					System.out.println("idDepot" + idDepot);
					System.out.println("capacityDepot" + capacityDepot);
					System.out.println("posCluster" + posCluster);
					
					if((posClusterOne != -1) && (posCluster != -1))
					{
						requestClusterOne = listClusters.get(posClusterOne).getRequestCluster();
						requestCluster = listClusters.get(posCluster).getRequestCluster();

						System.out.println("requestClusterOne" + requestClusterOne);
						System.out.println("requestCluster" + requestCluster);
						
						if(capacityDepot >= (requestCluster + requestClusterOne)) 
						{
							requestCluster += requestClusterOne;
							listClusters.get(posCluster).setRequestCluster(requestCluster);

							for (int i = 0; i < listClusters.get(posClusterOne).getItemsOfCluster().size(); i++)
								listClusters.get(posCluster).getItemsOfCluster().add(listClusters.get(posClusterOne).getItemsOfCluster().get(i));
					
							listCustomersToAssign.remove(Problem.getProblem().findPosCustomer(listCustomersToAssign, idCustomerOne));
							
							listClusters.remove(posClusterOne);
							change = true;
						}
						else
						{
							costMatrix.setItem(rcBestAll.getRow(), rcBestAll.getCol(), Double.POSITIVE_INFINITY);
							change = false;	
							
						/*	System.out.println();
							System.out.println("IDClusterOne" + listClusters.get(posClusterOne).getIDCluster());
							System.out.println("ElementClusterOne" + listClusters.get(posClusterOne).getItemsOfCluster().toString());
							System.out.println("requestClusterOne" + listClusters.get(posClusterTwo).getRequestCluster());*/
						}
						
						//PENDIENTE ENTRADA CON LISTCUSTOMER VACIA EN EL ISFULLDEPOT TRADICIONA
						if(isFullDepot(listClusters, requestCluster, capacityDepot, listCustomersToAssign.size()))
						{
							posCluster = findCluster(idDepot, listClusters);
							System.out.println("posCluster" + posCluster);
							
							if(!(listClusters.get(posCluster).getItemsOfCluster().isEmpty()))
								solution.getClusters().add(listClusters.remove(posCluster));
							else
								listClusters.remove(posCluster);

							listDepots.remove(posDepot);
							totalDepots--;
							change = true;
						}
					}
				}
			}
		}
		
		finish(listClusters, solution);
		
		if(!listClusters.isEmpty())
			for(int k = 0; k < listClusters.size(); k++)
				if(!(listClusters.get(k).getItemsOfCluster().isEmpty()))
					solution.getClusters().add(listClusters.get(k));

		return solution;
	}
	
	/*Método encargado de obtener el id del deposito con mayor capacidad de la lista*/
	private int getIDClusterWithMU(ArrayList<Depot> depots, ArrayList<Cluster> clusters){
		int idDepotMU = depots.get(0).getIDDepot();
		int posCluster = findCluster(idDepotMU, clusters);
		double requestCluster = clusters.get(posCluster).getRequestCluster();
		double maxCapacityDepot = Problem.getProblem().getTotalCapacityByDepot(depots.get(0)); 
		maxCapacityDepot -= requestCluster;
		int totalDepots = depots.size();

		double currentCapacityDepot; 

		for(int i = 1; i < totalDepots; i++)
		{
			posCluster = findCluster(depots.get(i).getIDDepot(), clusters);
			requestCluster = clusters.get(posCluster).getRequestCluster();
			currentCapacityDepot = Problem.getProblem().getTotalCapacityByDepot(depots.get(i));
			currentCapacityDepot -= requestCluster;
			
			if(maxCapacityDepot < currentCapacityDepot)
			{
				maxCapacityDepot = currentCapacityDepot;
				idDepotMU = depots.get(i).getIDDepot(); 
			}
		}

		return idDepotMU;
	}
	
	private void finish(ArrayList<Cluster> clusters, Solution solution){
		int posElement = -1;

		for(int i = 0; i < clusters.size(); i++)
		{
			System.out.println();
			System.out.println("IDCluster" + clusters.get(i).getIDCluster());
			System.out.println("posCluster" + clusters.get(i).getItemsOfCluster().toString());
			System.out.println("requestCluster" + clusters.get(i).getRequestCluster());
			
			posElement = Problem.getProblem().findPosCustomer(Problem.getProblem().getCustomers(), clusters.get(i).getIDCluster());

			if(posElement != -1)
			{
				for(int j = 0; j < clusters.get(i).getItemsOfCluster().size(); j++)
					solution.getUnassignedItems().add(clusters.get(i).getItemsOfCluster().get(j).intValue());

				clusters.remove(i);
			}
		}
	}
}